File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Listeners/onMove.java
package jp.aoichaan0513.A_TosoGame_Live.Listeners;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.ActionBarManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.OPGameManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.ArrayList;

public class onMove implements Listener {

    public static ArrayList<Player> zoneList = new ArrayList<>();

    @EventHandler(ignoreCancelled = true)
    public void onPlayerMove(PlayerMoveEvent e) {
        Player p = e.getPlayer();

        if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
            WorldConfig worldConfig = Main.getWorldConfig();

            if (isAllowArea(e.getTo(), worldConfig.getMapBorderConfig())) {
                e.setTo(e.getFrom());
                p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "マップから出ることは出来ません。");
                return;
            }

            if (MissionManager.isMission()) {
                if (MissionManager.getMission() == MissionManager.MissionType.MISSION_HUNTER_ZONE) {
                    if (isAllowArea(e.getTo(), worldConfig.getHunterZoneBorderConfig())) {
                        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p)) {
                            if (zoneList.contains(p)) {
                                zoneList.remove(p);
                            } else {
                                if (zoneList.size() < 3) {
                                    if (!zoneList.contains(p)) {
                                        zoneList.add(p);
                                    }
                                } else {
                                    e.setTo(e.getFrom());
                                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "3人以上は入れません。");
                                }
                            }
                        } else {
                            e.setTo(e.getFrom());
                            p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "現在入ることは出来ません。");
                        }
                        return;
                    }
                } else {
                    if (isAllowArea(e.getTo(), worldConfig.getHunterZoneBorderConfig())) {
                        if (!zoneList.contains(p)) {
                            e.setTo(e.getFrom());
                            p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "現在入ることは出来ません。");
                        } else {
                            zoneList.remove(p);
                        }
                        return;
                    }
                }
            } else {
                if (isAllowArea(e.getTo(), worldConfig.getHunterZoneBorderConfig())) {
                    if (!zoneList.contains(p)) {
                        e.setTo(e.getFrom());
                        p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "現在入ることは出来ません。");
                    } else {
                        zoneList.remove(p);
                    }
                    return;
                }
            }

            if (OPGameManager.getOPGame()) {
                if (OPGameManager.getDice()) {
                    if (OPGameManager.player != null) {
                        if (!p.getName().equals(OPGameManager.player.getName())) {
                            if (Main.playerList.contains(p)) {
                                if (isAllowArea(e.getTo(), worldConfig.getOPGameBorderConfig())) {
                                    e.setCancelled(true);
                                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "ここから外に出ることは出来ません。");
                                    new BukkitRunnable() {

                                        @Override
                                        public void run() {
                                            TosoGameAPI.teleport(p, worldConfig.getOPGameLocationConfig().getGOPLocations());
                                        }
                                    }.runTaskLater(Main.getInstance(), 1);
                                    return;
                                }
                            }
                        } else {
                            if (e.getFrom().getBlockX() != e.getTo().getBlockX() || e.getFrom().getBlockY() != e.getTo().getBlockY() || e.getFrom().getBlockZ() != e.getTo().getBlockZ())
                                e.setCancelled(true);
                        }
                    }
                }
            }

            if (!worldConfig.getGameConfig().getJump()) {
                if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_TUHO, p)) {
                    if (p.isSprinting() && !p.isOnGround() && e.getTo().getY() > e.getFrom().getY() + 0.2d) {
                        e.setTo(e.getFrom());

                        ActionBarManager.sendActionBar(p, ChatColor.YELLOW + "⚠ ダッシュジャンプは出来ません。");
                        return;
                    }
                }
            }
        }
    }

    public boolean isAllowArea(Location loc, WorldConfig.IBorderConfig borderConfig) {
        return borderConfig.isLocation(WorldConfig.BorderType.POINT_1) && borderConfig.isLocation(WorldConfig.BorderType.POINT_2) &&
                isAllowArea(loc, borderConfig.getLocation(WorldConfig.BorderType.POINT_1), borderConfig.getLocation(WorldConfig.BorderType.POINT_2));
    }

    public boolean isAllowArea(Location loc, Location loc1, Location loc2) {
        int x1 = Math.min(loc1.getBlockX(), loc2.getBlockX());
        int x2 = Math.max(loc1.getBlockX(), loc2.getBlockX());
        int y1 = Math.min(loc1.getBlockY(), loc2.getBlockY());
        int y2 = Math.max(loc1.getBlockY(), loc2.getBlockY());
        int z1 = Math.min(loc1.getBlockZ(), loc2.getBlockZ());
        int z2 = Math.max(loc1.getBlockZ(), loc2.getBlockZ());

        /*
        if (loc1.getBlockX() == loc.getBlockX()) {
            if (loc.getBlockZ() >= z1 && loc.getBlockZ() <= z2 || loc.getBlockZ() <= z1 && loc.getBlockZ() >= z2)
                if (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                    return true;
        } else if (loc1.getBlockZ() == loc.getBlockZ()) {
            if (loc.getBlockX() >= x1 && loc.getBlockX() <= x2 || loc.getBlockX() <= x1 && loc.getBlockX() >= x2)
                if (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                    return true;
        } else if (loc2.getBlockX() == loc.getBlockX()) {
            if (loc.getBlockZ() >= z1 && loc.getBlockZ() <= z2 || loc.getBlockZ() <= z1 && loc.getBlockZ() >= z2)
                if (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                    return true;
        } else if (loc2.getBlockZ() == loc.getBlockZ()) {
            if (loc.getBlockX() >= z1 && loc.getBlockX() <= z2 || loc.getBlockX() <= z1 && loc.getBlockX() >= z2)
                if (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                    return true;
        }
        return false;
        */

        return loc1.getBlockX() == loc.getBlockX() && (loc.getBlockZ() >= z1 && loc.getBlockZ() <= z2 || loc.getBlockZ() <= z1 && loc.getBlockZ() >= z2) && (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                || loc1.getBlockZ() == loc.getBlockZ() && (loc.getBlockX() >= x1 && loc.getBlockX() <= x2 || loc.getBlockX() <= x1 && loc.getBlockX() >= x2) && (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                || loc2.getBlockX() == loc.getBlockX() && (loc.getBlockZ() >= z1 && loc.getBlockZ() <= z2 || loc.getBlockZ() <= z1 && loc.getBlockZ() >= z2) && (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2)
                || loc2.getBlockZ() == loc.getBlockZ() && (loc.getBlockX() >= x1 && loc.getBlockX() <= x2 || loc.getBlockX() <= x1 && loc.getBlockX() >= x2) && (loc.getBlockY() >= y1 && loc.getBlockY() <= y2 || loc.getBlockY() <= y1 && loc.getBlockY() >= y2);
    }

    public Location getKnockbackLoc(Location fromLoc, Location toLoc, Location loc1, Location loc2) {
        if (loc1.getBlockX() == toLoc.getBlockX()) {
            if (toLoc.getBlockZ() >= loc1.getBlockZ() && toLoc.getBlockZ() <= loc2.getBlockZ() || toLoc.getBlockZ() <= loc1.getBlockZ() && toLoc.getBlockZ() >= loc2.getBlockZ()) {
                if (toLoc.getBlockX() >= fromLoc.getBlockX()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX() - 2, toLoc.getBlockY(), toLoc.getBlockZ(), toLoc.getYaw(), toLoc.getPitch());
                } else if (toLoc.getBlockX() <= fromLoc.getBlockX()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX() + 2, toLoc.getBlockY(), toLoc.getBlockZ(), toLoc.getYaw(), toLoc.getPitch());
                }
            }
        } else if (loc1.getBlockZ() == toLoc.getBlockZ()) {
            if (toLoc.getBlockX() >= loc1.getBlockX() && toLoc.getBlockX() <= loc2.getBlockX() || toLoc.getBlockX() <= loc1.getBlockX() && toLoc.getBlockX() >= loc2.getBlockX()) {
                if (toLoc.getBlockZ() >= fromLoc.getBlockZ()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX(), toLoc.getBlockY(), toLoc.getBlockZ() - 2, toLoc.getYaw(), toLoc.getPitch());
                } else if (toLoc.getBlockZ() <= fromLoc.getBlockZ()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX(), toLoc.getBlockY(), toLoc.getBlockZ() + 2, toLoc.getYaw(), toLoc.getPitch());
                }
            }
        } else if (loc2.getBlockX() == toLoc.getBlockX()) {
            if (toLoc.getBlockZ() >= loc1.getBlockZ() && toLoc.getBlockZ() <= loc2.getBlockZ() || toLoc.getBlockZ() <= loc1.getBlockZ() && toLoc.getBlockZ() >= loc2.getBlockZ()) {
                if (toLoc.getBlockX() >= fromLoc.getBlockX()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX() - 2, toLoc.getBlockY(), toLoc.getBlockZ(), toLoc.getYaw(), toLoc.getPitch());
                } else if (toLoc.getBlockX() <= fromLoc.getBlockX()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX() + 2, toLoc.getBlockY(), toLoc.getBlockZ(), toLoc.getYaw(), toLoc.getPitch());
                }
            }
        } else if (loc2.getBlockZ() == toLoc.getBlockZ()) {
            if (toLoc.getBlockX() >= loc1.getBlockX() && toLoc.getBlockX() <= loc2.getBlockX() || toLoc.getBlockX() <= loc1.getBlockX() && toLoc.getBlockX() >= loc2.getBlockX()) {
                if (toLoc.getBlockZ() >= fromLoc.getBlockZ()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX(), toLoc.getBlockY(), toLoc.getBlockZ() - 2, toLoc.getYaw(), toLoc.getPitch());
                } else if (toLoc.getBlockZ() <= fromLoc.getBlockZ()) {
                    return new Location(toLoc.getWorld(), toLoc.getBlockX(), toLoc.getBlockY(), toLoc.getBlockZ() + 2, toLoc.getYaw(), toLoc.getPitch());
                }
            }
        }
        return new Location(toLoc.getWorld(), toLoc.getBlockX(), toLoc.getBlockY(), toLoc.getBlockZ(), toLoc.getYaw(), toLoc.getPitch());
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Listeners/Minecraft/onMove.kt
package jp.aoichaan0513.A_TosoGame_Live.Listeners.Minecraft

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.ActionBarManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig.BorderType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig.IBorderConfig
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Mission.HunterZone
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.OPGame.OPGameManager
import jp.aoichaan0513.A_TosoGame_Live.Utils.isAdminTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isJailTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isPlayerGroup
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent


class onMove : Listener {

    @EventHandler(ignoreCancelled = true)
    fun onPlayerMove(e: PlayerMoveEvent) {
        val p = e.player
        val from = e.from
        val to = e.to!!

        if (p.isAdminTeam) return

        val worldConfig = Main.worldConfig
        val mapBorder = worldConfig.mapBorderConfig
        val hunterZoneBorder = worldConfig.hunterZoneBorderConfig

        if (isBorderAttack(to, worldConfig.mapBorderConfig)) {
            e.setTo(from)
            p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}マップから出ることはできません。")
        }


        val fromIsIn = isJoinedArea(from, hunterZoneBorder)
        val toIsIn = isJoinedArea(to, hunterZoneBorder)

        if (fromIsIn && !toIsIn) {
            // エリアから出たとき
            if (p.isPlayerGroup) {
                HunterZone.removeJoinedSet(p)
            } else {
                e.setTo(from)
            }
        } else if (!fromIsIn && toIsIn) {
            if (p.isPlayerGroup) {
                if (MissionManager.isMissions && MissionManager.isMission(MissionManager.MissionState.HUNTER_ZONE)) {
                    if (HunterZone.joinedSetCount < 3) {
                        HunterZone.addJoinedSet(p)
                    } else {
                        e.setTo(from)
                        p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}4人以上は入れません。")
                    }
                } else {
                    e.setTo(from)
                    p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}現在入ることはできません。")
                }
            } else {
                e.setTo(from)
                p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}ハンター、通報部隊は入ることができません。")
            }
        }


        if (OPGameManager.opGameState != OPGameManager.OPGameState.NONE) {
            val player = OPGameManager.player
            if (player != null && p.isPlayerGroup) {
                if (OPGameManager.runnedSet.contains(p.uniqueId)) {
                    if (player.uniqueId == p.uniqueId)
                        if (from.blockX != to.blockX || from.blockY != to.blockY || from.blockZ != to.blockZ)
                            e.isCancelled = true
                } else {
                    if (isBorderAttack(to, worldConfig.opGameBorderConfig)) {
                        e.isCancelled = true
                        p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}ここから外に出ることはできません。")
                        TosoGameAPI.teleport(p, worldConfig.opGameLocationConfig.gOPLocations.values)
                    }
                }
            }
        }

        if (!worldConfig.gameConfig.jump) {
            if (p.isJailTeam || !p.isSprinting
                    || p.isOnGround || to.y < from.y + 0.3) return
            e.setTo(from)
            ActionBarManager.send(p, "${ChatColor.YELLOW}⚠ ダッシュジャンプはできません。")
        }
    }

    fun isBorderAttack(loc: Location, borderConfig: IBorderConfig): Boolean {
        if (!borderConfig.isLocation(BorderType.POINT_1) || !borderConfig.isLocation(BorderType.POINT_2)) return false

        val loc1 = borderConfig.getLocation(BorderType.POINT_1)
        val loc2 = borderConfig.getLocation(BorderType.POINT_2)

        return isBorderAttack(loc, loc1, loc2)
    }

    fun isBorderAttack(loc: Location, loc1: Location, loc2: Location): Boolean {
        val x1 = loc1.blockX.coerceAtMost(loc2.blockX)
        val x2 = loc1.blockX.coerceAtLeast(loc2.blockX)
        val y1 = loc1.blockY.coerceAtMost(loc2.blockY) - 1
        val y2 = loc1.blockY.coerceAtLeast(loc2.blockY) + 1
        val z1 = loc1.blockZ.coerceAtMost(loc2.blockZ)
        val z2 = loc1.blockZ.coerceAtLeast(loc2.blockZ)

        val isXAllowArea = loc.blockX in x1..x2
        val isYAllowArea = loc.blockY in y1..y2
        val isZAllowArea = loc.blockZ in z1..z2

        return ((x1 == loc.blockX || x2 == loc.blockX) && isYAllowArea && isZAllowArea)
                || ((y1 == loc.blockY || y2 == loc.blockY) && isXAllowArea && isZAllowArea)
                || ((z1 == loc.blockZ || z2 == loc.blockZ) && isXAllowArea && isYAllowArea)
    }

    fun isJoinedArea(loc: Location, borderConfig: IBorderConfig, isOutside: Boolean = false): Boolean {
        if (!borderConfig.isLocation(BorderType.POINT_1) || !borderConfig.isLocation(BorderType.POINT_2)) return false

        val loc1 = borderConfig.getLocation(BorderType.POINT_1)
        val loc2 = borderConfig.getLocation(BorderType.POINT_2)

        return isJoinedArea(loc, loc1, loc2, isOutside)
    }

    fun isJoinedArea(loc: Location, loc1: Location, loc2: Location, isOutside: Boolean = false): Boolean {
        val x1 = loc1.blockX.coerceAtMost(loc2.blockX)
        val x2 = loc1.blockX.coerceAtLeast(loc2.blockX)
        val y1 = loc1.blockY.coerceAtMost(loc2.blockY) - 1
        val y2 = loc1.blockY.coerceAtLeast(loc2.blockY) + 1
        val z1 = loc1.blockZ.coerceAtMost(loc2.blockZ)
        val z2 = loc1.blockZ.coerceAtLeast(loc2.blockZ)

        val isXJoinedArea = loc.blockX in if (isOutside) (x1 + 1) until x2 else x1..x2
        val isYJoinedArea = loc.blockY in y1..y2
        val isZJoinedArea = loc.blockZ in if (isOutside) (z1 + 1) until z2 else z1..z2

        return isXJoinedArea && isYJoinedArea && isZJoinedArea
    }
}